### PR TITLE
Homework23a

### PR DESCRIPTION
**Pull request by trtestpro**

### CODE REVIEW

File: src/test/java/pages/LoginPage.java [hunk 1]
```diff
@@ -39,6 +48,9 @@ public void providePassword(String password) {
 
     public void logIn(){
         provideEmail("terrence.regis@gmail.com");
+
```
Comment (Berichicko) [inline src/test/java/pages/LoginPage.java:51]: The code is not compiled! Please finish the homework!You got this problem, please correct it. 

image

For example

```
package pages;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.FindBy;

public class LoginPage extends BasePage {

    //locators
    @FindBy(css = "[type='submit']")
    private WebElement submitButtonLocator;
    @FindBy(css = "[type='email']")
    private WebElement emailField;
    @FindBy(css = "[type='password']")
    private WebElement passwordField;


    public LoginPage(WebDriver givenDriver) {
        super(givenDriver);
    }


    public void clickSubmitBtn() {
        submitButtonLocator.click();
    }

    public void provideEmail(String email) {
        emailField.sendKeys(email);
    }

    public void providePassword(String password) {
        passwordField.sendKeys(password);

    }

    public void logIn() {
        provideEmail("terrence.regis@gmail.com");
        providePassword("te$t$tudent");
        clickSubmitBtn();

    }

}
```

File: src/test/java/BaseTest.java [hunk 2]
```diff
@@ -38,6 +39,18 @@ static void setupClass() {
...
     }
 
+    public WebDriver driver = null;
+    public String url = null;
+    public WebDriverWait wait = null;
+
+    public FluentWait fluentWait = null;
+    public Actions actions = null;
+    public ThreadLocal<WebDriver> threadLocal = null;
+
```
Comment (Berichicko) [inline src/test/java/BaseTest.java:49]: **Please, remove code redundancies**

Don't repeat yourself! ( In Java, it means don't write the same code repeatedly. )

image

**Comment by trtestpro** [comment]
> _ No description provided. _

**Comment by trtestpro** [comment]
The first comment is that the code does not compile and I have a list of errors. 
Error 1 - there is no constructor in pages BasePage.  This is my BasePage code I commited - that ran compiled and ran fine w/o issues. 

public class BasePage {
    protected static WebDriver driver;
    protected static WebDriverWait wait;
    protected static Actions actions;

    public BasePage( WebDriver givenDriver) {
        driver = givenDriver;
        wait = new WebDriverWait(driver, Duration.ofSeconds(10));
        actions = new Actions(driver);
        PageFactory.initElements(driver, this);
    }

    public void click(By locator) {
        wait.until(ExpectedConditions.elementToBeClickable(locator));
    }

    public static WebElement findElement(By locator) {
        return wait.until(ExpectedConditions.visibilityOfElementLocated(locator));
    }

    public void contextClick(By locator) {
        actions.contextClick(findElement(locator)).perform();
    }

    public void doubleClick (By locator) {
        actions.doubleClick(findElement(locator)).perform();
    }
}

Error 2 - Said Duplicate Classes Loginpage; plus I was told not to repeat where it was hsowed I defined the Fluentwait twice.  If you look at the code I pasted above, it does not even have a mention of Fluentwait.  Furthermore, see below for my checked in code for LoginPage - no duplicates - unlesss I am missing something.  

public class LoginPage extends BasePage{

    //Locators
     @FindBy(css = "[type='submit']" )
    private WebElement submitButtonLocator;
    @FindBy(css="[type='email']")
    private WebElement emailField;
    @FindBy(css="[type='password']")
    private WebElement passwordField;


    public LoginPage( WebDriver givenDriver) {
        super(givenDriver);


    }

    public void clickSubmitBtn() {
        submitButtonLocator.click();
    }

    public void provideEmail(String email) {
        emailField.sendKeys(email);
    }

    public void providePassword(String password) {
        passwordField.sendKeys(password);

    }

    public void logIn(){
        provideEmail("terrence.regis@gmail.com");
        providePassword("te$t$tudent");
        clickSubmitBtn();
    }


I am also including the code for HomePage below.  
public class HomePage extends BasePage {
    private static String playlistName = "Summer Songs";
    private  static By playlistElementLocator = By.xpath("//a[text() = '"+playlistName+"']");

    @FindBy(css="img.avatar")
    private WebElement userAvatarIcon;

    @FindBy(css=".playlist:nth-child(3)")
    private static WebElement playlistLocator;

    @FindBy(css="input[name='name']")
    private static WebElement playlistNameLocator;

    public HomePage( WebDriver givenDriver) {
        super(givenDriver);
    }

    public WebElement getUserAvatar () {
        return userAvatarIcon;
    }


}

I am not sure if it's because I am behind in homework submission that things are getting screwed up; but there is clearly a difference between what I am submitting and what is being corrected.  i am not sure if its getting messed up in the merge or somewhere else.        

Error(Issue) 3 - Whenever I checkout origin/main - I am not getting all the homeworks in the branch; I have to resort to pulling my lastest homework branch to complete the work.  This started as off HW23.  

Screenshot 2023-01-31 at 11 28 34 AM

Before I submitted, I tested and fixed all issues.  I also sent you the test report that showed all test passed.   I know that the merge is an important part also that's why I need your help to see how I can get back in sync so my homework could be appropriately evaluated.  Thank you.

**Comment by trtestpro** [comment]
One quick correction as far s the duplicate for FluentWait.  It's in the BaseTest class, I accidentally said BasePage - below is the BaseTest class on my local FluentWait is define/declared once.

public class BaseTest {
    public static WebDriver driver = null;
    public static String url = null;
    public static WebDriverWait wait = null;
    public static FluentWait fluentWait = null;


    @BeforeSuite
    static void setupClass() {
        WebDriverManager.chromedriver().setup();
    }

    @BeforeMethod
    @Parameters({"BaseURL"})
    public static void launchBrowser(String BaseURL) {
        LoginTests.driver = new ChromeDriver();
//        LoginTests.driver.manage().timeouts().implicitlyWait(Duration.ofSeconds(10));
        url = BaseURL;
        driver.get(url);
        wait = new WebDriverWait(LoginTests.driver, Duration.ofSeconds(20));
    }

    @AfterMethod
    public static void closeBrowser(){
        LoginTests.driver.quit();
    }

//    protected static void navigateToPage() {
//        String url = "https://bbb.testpro.io/";
//        driver.get(url);
//    }

    public static void login(String email, String password) {
        provideEmail(email);
        providePassword(password);
        clickSubmit();
    }

    public static void clickSubmit() {
        WebElement submitButton = wait.until(ExpectedConditions.elementToBeClickable(By.cssSelector("button[type='submit']")));
        submitButton.click();
    }

    public static void providePassword(String password) {
        WebElement passwordField = driver.findElement(By.cssSelector("[type='password']"));
        wait.until(ExpectedConditions.elementToBeClickable(passwordField));// use this when method only take WebElement

        passwordField.clear();
        passwordField.sendKeys(password);
    }

    public static void provideEmail(String email) {
        WebElement emailField = wait.until(ExpectedConditions.elementToBeClickable(By.cssSelector("[type='email']")));
        emailField.clear();
        emailField.sendKeys(email);
    }

    public static void clickSaveButton() {
        WebElement saveButton = driver.findElement(By.cssSelector("button.btn-submit"));
        saveButton.click();
    }

    public static void provideProfileName(String randomName) {
        WebElement profileName = driver.findElement(By.cssSelector("[name='name']"));
        profileName.clear();
        profileName.sendKeys(randomName);
    }

    public static void provideCurrentPassword(String password) {
        WebElement currentPassword = driver.findElement(By.cssSelector("[name='current_password']"));
        currentPassword.clear();
        currentPassword.sendKeys(password);
    }

    public static String generateRandomName() {
        return UUID.randomUUID().toString().replace("-", "");//
    }

    public static void clickAvatarIcon() {
        WebElement avatarIcon = driver.findElement(By.cssSelector("img.avatar"));
        avatarIcon.click();

    }

    @DataProvider(name="incorrectLoginProviders")
    public static Object[][] getDataFromDataproviders() {

        return new Object[][] {
                {"invalid@email.com", "invalidPass"},
                {"demo@mail.com", "invalid"},
                {"", ""}
        };
    }

**Pull request closed without merging**